expression.go

```diff
@@ -0,0 +1,195 @@
+	case *proto.Expression_Selection:
+		return FieldReferenceFromProto(et.Selection, baseSchema, ext)
+	case *proto.Expression_ScalarFunction_:
+		var err error
+		args := make([]FuncArg, len(et.ScalarFunction.Arguments))
+		for i, a := range et.ScalarFunction.Arguments {
+			if args[i], err = FuncArgFromProto(a, baseSchema, ext); err != nil {
+				return nil, err
+			}
+		}
+
+		id, ok := ext.DecodeFunc(et.ScalarFunction.FunctionReference)
+		if !ok {
+			return nil, ErrNotFound
+		}
+
+		return &ScalarFunction{
+			FuncRef:    et.ScalarFunction.FunctionReference,
+			ID:         id,
+			Args:       args,
+			Options:    et.ScalarFunction.Options,
+			OutputType: TypeFromProto(et.ScalarFunction.OutputType),
+		}, nil
+	case *proto.Expression_WindowFunction_:
+		var err error
+		args := make([]FuncArg, len(et.WindowFunction.Arguments))
+		for i, a := range et.WindowFunction.Arguments {
+			if args[i], err = FuncArgFromProto(a, baseSchema, ext); err != nil {
+				return nil, err
+			}
+		}
+
+		parts := make([]Expression, len(et.WindowFunction.Partitions))
+		for i, p := range et.WindowFunction.Partitions {
+			if parts[i], err = ExprFromProto(p, baseSchema, ext); err != nil {
+				return nil, err
+			}
+		}
+
+		sorts := make([]SortField, len(et.WindowFunction.Sorts))
+		for i, s := range et.WindowFunction.Sorts {
+			if sorts[i], err = SortFieldFromProto(s, baseSchema, ext); err != nil {
+				return nil, err
+			}
+		}
+
+		id, ok := ext.DecodeFunc(et.WindowFunction.FunctionReference)
+		if !ok {
+			return nil, ErrNotFound
+		}
+
+		return &WindowFunction{
+			FuncRef:    et.WindowFunction.FunctionReference,
+			ID:         id,
+			Args:       args,
+			Options:    et.WindowFunction.Options,
+			OutputType: TypeFromProto(et.WindowFunction.OutputType),
+			Phase:      et.WindowFunction.Phase,
+			Invocation: et.WindowFunction.Invocation,
+			Partitions: parts,
+			Sorts:      sorts,
+			LowerBound: BoundFromProto(et.WindowFunction.LowerBound),
+			UpperBound: BoundFromProto(et.WindowFunction.UpperBound),
+		}, nil
+	case *proto.Expression_IfThen_:
+	case *proto.Expression_SwitchExpression_:
+	case *proto.Expression_SingularOrList_:
+	case *proto.Expression_MultiOrList_:
+	case *proto.Expression_Cast_:
+	case *proto.Expression_Nested_:
+	case *proto.Expression_Enum_:
+		return nil, fmt.Errorf("%w: deprecated", ErrNotImplemented)
+	case *proto.Expression_Subquery_:
+	}
+
+	return nil, ErrNotImplemented
+}
+
+// Expression can be one of many different things as a generalized
+// expression. It could be:
+//
+//  - A literal
+//  - A Field Reference Selection
+//  - A Scalar Function expression
+//  - A Window Function expression
+//  - An If-Then statement
+//  - A Switch Expression
+//  - A Singular Or List
+//  - A Multiple Or List
+//  - A Cast expression
+//  - A Subquery
+//  - A Nested expression
+type Expression interface {
+	FuncArg
+	RootRefType
+
+	GetType() Type
+	ToProto() *proto.Expression
+	Equals(Expression) bool
```

Comment:
I think we'll need a general way of walking through children, no? Like `Walk(Visitor) Expression`

Reply:
I'll follow the pattern from optgen that you linked earlier where the visiting could be used to generate a replacement tree if desired etc.

Reply:
added